### PR TITLE
Rerun `configure` less often

### LINKED ISSUE #32119: Makefile: If build/make/Makefile.in changes, only run ./config.status, not ... --recheck
<div id="comment:0"></div>

Component: **build**

_Issue created by migration from https://trac.sagemath.org/ticket/32119_

### PR DESCRIPTION
**Pull request by mkoeppe**
There's no need to re-run `configure` if only template files such as `build/make/Makefile.in` (see `AC_CONFIG_FILES` in `configure.ac`) change: It suffices to run `config.status` in this case.

Fixes #32119

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [ ] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -47,10 +47,21 @@ CONFIG_FILES = build/make/Makefile src/bin/sage-env-config build/bin/sage-build-
...
-# Otherwise, run configure with argument $PREREQ_OPTIONS.
+# If configure was not run before, complain.
+# If configure is newer than the files it generated (we test build/make/Makefile),
+#    we regenerate config.status by running the "config.status --recheck".
+# Either way we regenerate the generated files by calling "config.status".
 build/make/Makefile: configure $(SPKG_COLLECT_FILES) $(CONFIG_FILES:%=%.in)
-	$(MAKE) reconfigure
+	@if [ -x config.status ]; then		\
+	    case '$?' in			\
+		*configure*)			\
```
Comment (kwankyu) [inline Makefile]: What is the previous command for `'$?'`? I have no idea...

Comment (mkoeppe) [reply Makefile]: This is not a shell variable (that would be `$$?` within a make recipe), it's an "automatic variable" of the Makefile rule: https://www.gnu.org/software/make/manual/html_node/Automatic-Variables.html
-  `$?` The names of all the prerequisites that are newer than the target, with spaces between them. If the target does not exist, all prerequisites will be included.

Comment (kwankyu) [reply Makefile]: Thanks. That fills a big gap in my rudimentary knowledge on `Makefile`.

Comment (kwankyu) [reply Makefile]: If a file in `SPKG_COLLECT_FILES` changes, no need to run "config.status --recheck"?

Comment (mkoeppe) [reply Makefile]: In this case, `configure` will first be regenerated (that's another Makefile rule) and thus `configure` will be part of `$?`

Comment (kwankyu) [reply Makefile]: I see. It is 
```
configure: bootstrap src/doc/bootstrap configure.ac src/bin/sage-version.sh m4/*.m4 build/pkgs/*/spkg-configure.m4 build/pkgs/*/type build/pkgs/*/install-requires.txt build/pkgs/*/package-version.txt build/pkgs/*/distros/*.txt  
```

Then I wonder why `SPKG_COLLECT_FILES` is not used here as a prerequisite, though I see that the list contains different files.

Comment (mkoeppe) [reply Makefile]: That's a very good question. I'll investigate.

Comment (mkoeppe) [reply Makefile]: I've reworked it now

File: Makefile [hunk 2]
```diff
@@ -324,7 +336,14 @@ ptestoptionallong-nodoc:
...
-configure: bootstrap src/doc/bootstrap configure.ac src/bin/sage-version.sh m4/*.m4 build/pkgs/*/spkg-configure.m4 build/pkgs/*/type build/pkgs/*/install-requires.txt build/pkgs/*/package-version.txt build/pkgs/*/distros/*.txt
+# The 'configure' script is just one of the files generated by 'bootstrap'.
+# CONFIGURE_DEPENDENCIES is the list of files that influence the generation of 'configure'.
+CONFIGURE_DEPENDENCIES=configure.ac src/bin/sage-version.sh m4/*.m4 build/pkgs/*/spkg-configure.m4 \
+	build/pkgs/*/type build/pkgs/*/SPKG.rst build/pkgs/*/checksums.ini build/pkgs/*/requirements.txt build/pkgs/*/install-requires.txt build/pkgs/*/package-version.txt build/pkgs/*/spkg-install build/pkgs/*/spkg-install.in
+# SPKG_INFO_DEPENDENCIES is the list of files that influence the run of 'sage-spkg-info' and hence
+# the generation of the files generated in 'src/doc' by 'src/doc/bootstrap'.
+SPKG_INFO_DEPENDENCIES=build/pkgs/*/type build/pkgs/*/SPKG.rst build/pkgs/*/requirements.txt build/pkgs/*/install-requires.txt build/pkgs/*/package-version.txt build/pkgs/*/distros/*.txt
+configure: bootstrap src/doc/bootstrap $(CONFIGURE_DEPENDENCIES) $(SPKG_INFO_DEPENDENCIES)
 	./bootstrap -d
```
Comment (kwankyu) [inline Makefile:359]: To wrap the long list:
```diff
--- a/Makefile
+++ b/Makefile
@@ -338,11 +338,17 @@ ptestoptionallong-nodoc:
 
 # The 'configure' script is just one of the files generated by 'bootstrap'.
 # CONFIGURE_DEPENDENCIES is the list of files that influence the generation of 'configure'.
-CONFIGURE_DEPENDENCIES=configure.ac src/bin/sage-version.sh m4/*.m4 build/pkgs/*/spkg-configure.m4 \
-       build/pkgs/*/type build/pkgs/*/SPKG.rst build/pkgs/*/checksums.ini build/pkgs/*/requirements.txt build/pkgs/*/install-requires.txt build/pkgs/*/package-version.txt build/pkgs/*/spkg-install build/pkgs/*/spkg-install.in
+CONFIGURE_DEPENDENCIES=configure.ac src/bin/sage-version.sh m4/*.m4 \
+       build/pkgs/*/spkg-configure.m4 build/pkgs/*/spkg-install build/pkgs/*/spkg-install.in build/pkgs/*/checksums.ini \
+       build/pkgs/*/type build/pkgs/*/SPKG.rst build/pkgs/*/requirements.txt \
+       build/pkgs/*/install-requires.txt build/pkgs/*/package-version.txt
+
 # SPKG_INFO_DEPENDENCIES is the list of files that influence the run of 'sage-spkg-info' and hence
 # the generation of the files generated in 'src/doc' by 'src/doc/bootstrap'.
-SPKG_INFO_DEPENDENCIES=build/pkgs/*/type build/pkgs/*/SPKG.rst build/pkgs/*/requirements.txt build/pkgs/*/install-requires.txt build/pkgs/*/package-version.txt build/pkgs/*/distros/*.txt
+SPKG_INFO_DEPENDENCIES=build/pkgs/*/distros/*.txt \
+       build/pkgs/*/type build/pkgs/*/SPKG.rst build/pkgs/*/requirements.txt \
+       build/pkgs/*/install-requires.txt build/pkgs/*/package-version.txt
+
 configure: bootstrap src/doc/bootstrap $(CONFIGURE_DEPENDENCIES) $(SPKG_INFO_DEPENDENCIES)
        ./bootstrap -d
```                       
and compare. Doesn't this ` build/pkgs/*/distros/*.txt` influence the generation of `configure`?

Comment (mkoeppe) [reply Makefile:359]: > Doesn't this ` build/pkgs/*/distros/*.txt` influence the generation of `configure`?

No. `configure` only uses that at its runtime when it calls `sage-get-system-packages`

Comment (kwankyu) [reply Makefile:359]: Would you do the wrapping I suggested?

Otherwise it looks good to me.

Comment (mkoeppe) [reply Makefile:359]: I've reformatted in a similar way in 663c222

Comment (kwankyu) [reply Makefile:359]: I wanted it to be easy to compare `CONFIGURE_DEPENDENCIES` and `SPKG_INFO_DEPENDENCIES` by reordering items. But that is a minor thing.

Better than before. Thanks.

**Review by kwankyu — APPROVED** [review]
LGTM.

**Comment by mkoeppe** [comment]
Thank you!

**Comment by kwankyu** [comment]
I wonder if the failed test in B&T is related...

**Pull request merged**